oio/xcute/orchestrator.py

```diff
@@ -262,50 +262,94 @@ def dispatch_tasks(self, job_id, job_type, job_info, job):
+                    break
+                sleep(1)
+
+            if exc is not None and not self.running:
                 break
+            tasks.clear()
         else:
-            sent = True
-            if tasks:
-                sent = self.dispatch_tasks_batch(
-                    beanstalkd_workers,
-                    job_id, job_type, job_config, tasks)
-            if sent:
-                job_status, exc = self.handle_backend_errors(
+            # Make sure that the sent tasks will be saved
+            # before being processed
+            sent = False
+            while not sent:
+                (job_status, old_last_sent), exc = self.handle_backend_errors(
                     self.backend.update_tasks_sent, job_id, tasks.keys(),
                     all_tasks_sent=True)
-                if exc is None:
+                if exc is not None:
+                    self.logger.warn(
+                        '[job_id=%s] Job could not update '
+                        'the sent tasks: %s', job_id, exc)
+                    break
+                if tasks:
+                    sent = self.dispatch_tasks_batch(
+                        beanstalkd_workers, job_id, job_type, job_config,
+                        tasks)
+                else:
+                    sent = True
+                if not sent:
+                    self.logger.warn(
+                        '[job_id=%s] Job aborting the last sent tasks', job_id)
+                    job_status, exc = self.handle_backend_errors(
+                        self.backend.abort_tasks_sent, job_id, tasks.keys(),
+                        old_last_sent)
+                    if exc is not None:
+                        self.logger.warn(
+                            '[job_id=%s] Job could not abort '
+                            'the last sent tasks: %s', job_id, exc)
+                        break
+                else:
                     if job_status == 'FINISHED':
                         self.logger.info('Job %s is finished', job_id)
 
                     self.logger.info(
                         'Finished dispatching job (job_id=%s)', job_id)
                     return
-                else:
-                    self.logger.warn(
-                        '[job_id=%s] Job has not been updated '
-                        'with the last sent tasks: %s', job_id, exc)
+                if job_status == 'PAUSED':
+                    self.logger.info('Job %s is paused', job_id)
+                    return
+
+                if not self.running:
+                    break
+                sleep(1)
+
+        self.logger.warn(
+            '[job_id=%s] Job was stopped before it was finished', job_id)
 
         _, exc = self.handle_backend_errors(self.backend.free, job_id)
         if exc is not None:
             self.logger.warn(
-                '[job_id=%s] Job has not been freed: %s', job_id, exc)
+                '[job_id=%s] Job not be freed: %s', job_id, exc)
```

Comment:
"Job has not been freed" was correct. "Job not freed" would be correct. "Job not be freed" sounds funny.